### ISSUE TITLE
Attrs are lost in mathematical computation

### ISSUE THREAD

**Issue by fujiisoup** [pydata/xarray#1271]
Related to #138

Why is `keep_attrs` option in reduce method set to FALSE by default?
I feel it is more natural to keep all the attrs after some computation that returns xr.DaraArray
Such as `data*100`.
(By it is not possible to set this option TRUE when using an operator.)
Is it an option to move this option to __init__ method, in case of TRUE all the attrs are tracked after computations of the object and also the object generated from this object?

**Comment by shoyer** [comment]
Historically, we dropped attributes in arithmetic because attributes are often used for units and we didn't want to do computation that results in the wrong units. Blinding propagating metadata can lead to it ending up in the wrong place.

Also, if you are combining multiple DataArray objects with different `attrs`, there are a number of options for combining them, and it wasn't obvious which is the right strategy.

But in practice, this is something that a lot of people want (better to have stale metadata than no metadata at all), so maybe I made the wrong choice here. At the very least, I would be comfortable with an [option](http://xarray.pydata.org/en/stable/generated/xarray.set_options.html) to set `keep_attrs=True` for every operation and/or to specify a merge strategy for combining `attrs` in arithmetic.

**Comment by fujiisoup** [comment]
Thank you for the information.
I agree that we should not strongly rely on attrs. Unit may change in arithmetic.

My sense is closest to the option 3 in #131.
Some attrs should be tracked and other should be dropped.
In the present stage, the most possible option is to add `keep_attrs` in `xr.set_options`?
Or any destructive change will be an option?
Such as to divide attrs into two kinds, `units` and `metadata`.

**Comment by fmaussion** [comment]
> divide attrs into two kinds, units and metadata

Coordinate reference systems (crs) are an example of attribute which is always valid after selection, arithmetic, or even reduction since they apply to the coordinates of the DataArray. A global option to keep certain attrs (like ``crs``) would be very useful to downstream libraries like salem.

**Comment by dopplershift** [comment]
Could the setting not be module-wide, but be set on `DataArray` instances themselves? That setting could then be inherited when new ones are created from arithmetic operations.

I'd like to have the option of specifying specific attributes that should be kept, or maybe dropped. It will be *really* annoying to need to keep copying the `grid_mapping` attribute.

**Comment by shoyer** [comment]
> Coordinate reference systems (crs) are an example of attribute which is always valid after selection, arithmetic, or even reduction since they apply to the coordinates of the DataArray.

I wonder if it might make sense to represent `crs` as a (scalar) coordinate, which would already be propagated in all unambiguous cases by xarray ops.

The main reason why this is annoying is that `ds.crs` would return a 0-dimensional DataArray, so you would need to write `ds.crs.item()` to access it directly (e.g., if you want to call a method on it). 

Though you could actually already use the accessor interface to override this:
```
@xr.register_dataset_accessor('crs')
@xr.register_dataarray_accessor('crs')
def make_crs(xarray_obj):
     return xarray_obj.coords['crs'].item()
```
Usage:
```
In [18]: ds = xr.Dataset({'x': 1}, coords={'foo': 'bar', 'crs': 'hello, my lovely unwrapped object'})

In [19]: ds
Out[19]:
<xarray.Dataset>
Dimensions:  ()
Coordinates:
    foo      <U3 'bar'
    crs      <U33 'hello, my lovely unwrapped object'
Data variables:
    x        int64 1

In [20]: ds.crs
Out[20]: 'hello, my lovely unwrapped object'

In [21]: ds.foo.crs
Out[21]: 'hello, my lovely unwrapped object'

In [22]: (ds + 10).crs
Out[22]: 'hello, my lovely unwrapped object'
```

> Could the setting not be module-wide, but be set on DataArray instances themselves? That setting could then be inherited when new ones are created from arithmetic operations.

I would strong prefer to avoid making the xarray data model more complex by adding another type of metadata (e.g., in addition to `dims`, `coords`, `attrs` and `name` on DataArray). Whitelisting specific `attrs` as ones that should be preserved in application code via `set_options` (or encouraging users to subclass DataArray in a "safe" way) is preferable in that regard. This is related to the fuller "hook" system for customized attribute handling that I imagined over in #988.

**Comment by fmaussion** [comment]
> I wonder if it might make sense to represent crs as a (scalar) coordinate, which would already be propagated in all unambiguous cases by xarray ops.

This looks like a brilliant idea, thanks!

**Comment by fujiisoup** [comment]
> I would strong prefer to avoid making the xarray data model more complex by adding another type of metadata (e.g., in addition to dims, coords, attrs and name on DataArray). Whitelisting specific attrs as ones that should be preserved in application code via set_options (or encouraging users to subclass DataArray in a "safe" way) is preferable in that regard.

Understood.
I close this issue.
I guess discussions for how this option would be implemented could be continued in #988.
